### ISSUE TITLE
Incorrect type of return value

### ISSUE THREAD

**Issue by decaz** [ronf/asyncssh#440]
Method `SFTPClient.open` should return value of type `SFTPClientFile` instead of `SFTPFileProtocol`:
https://github.com/ronf/asyncssh/blob/215dbf63fd82270716814de63e045c512d0e5b72/asyncssh/sftp.py#L4165-L4180

**Comment by ronf** [comment]
Since there are some additional methods on SFTPClientFile which are not currently part of SFTPFileProtocol, I agree that this would be a good change. Unfortunately, it triggers a problem in mypy. After making this change, I get the following errors related to the \@async_context_manager decorator around that function:

```python
asyncssh/sftp.py:3657: error: Argument 1 to "_begin_copy" of "SFTPClient" has incompatible type "SFTPClient"; expected "_SFTPFSProtocol"
asyncssh/sftp.py:3657: note: Following member(s) of "SFTPClient" have conflicts:
asyncssh/sftp.py:3657: note:     Expected:
asyncssh/sftp.py:3657: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPFileProtocol]
asyncssh/sftp.py:3657: note:     Got:
asyncssh/sftp.py:3657: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPClientFile]
asyncssh/sftp.py:3762: error: Argument 2 to "_begin_copy" of "SFTPClient" has incompatible type "SFTPClient"; expected "_SFTPFSProtocol"
asyncssh/sftp.py:3762: note: Following member(s) of "SFTPClient" have conflicts:
asyncssh/sftp.py:3762: note:     Expected:
asyncssh/sftp.py:3762: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPFileProtocol]
asyncssh/sftp.py:3762: note:     Got:
asyncssh/sftp.py:3762: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPClientFile]
asyncssh/sftp.py:3891: error: Argument 1 to "_begin_copy" of "SFTPClient" has incompatible type "SFTPClient"; expected "_SFTPFSProtocol"
asyncssh/sftp.py:3891: note: Following member(s) of "SFTPClient" have conflicts:
asyncssh/sftp.py:3891: note:     Expected:
asyncssh/sftp.py:3891: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPFileProtocol]
asyncssh/sftp.py:3891: note:     Got:
asyncssh/sftp.py:3891: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPClientFile]
asyncssh/sftp.py:3915: error: Argument 2 to "_begin_copy" of "SFTPClient" has incompatible type "SFTPClient"; expected "_SFTPFSProtocol"
asyncssh/sftp.py:3915: note: Following member(s) of "SFTPClient" have conflicts:
asyncssh/sftp.py:3915: note:     Expected:
asyncssh/sftp.py:3915: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPFileProtocol]
asyncssh/sftp.py:3915: note:     Got:
asyncssh/sftp.py:3915: note:         def open(*Any, **Any) -> _ACMWrapper[SFTPClientFile]
```

Basically, making this change makes `SFTPClient` no longer compatible with the `_SFTPFSProtocol`, which is needed for the various SFTP copy functions that want to be able to accept either an `SFTPClient` or a `LocalFS`.

I'm not quite sure how to get around this problem, so I eventually just settled for having the `open` method of both `SFTPClient` and `LocalFS` return the protocol class. This is far from ideal, though, so any suggestions are welcome.

**Comment by ronf** [comment]
Good news on this. I thought about it a bit more and found the issue. I needed to mark one of the type variables used in the async_context_manager as "covariant". The following changes seem to avoid the mypy errors I saw:

```patch
diff --git a/asyncssh/misc.py b/asyncssh/misc.py
index 7f08a1c..306deca 100644
--- a/asyncssh/misc.py
+++ b/asyncssh/misc.py
@@ -258,7 +258,7 @@ def parse_time_interval(value: str) -> float:
     return _parse_units(value, _time_units, 'time interval')


-_ACM = TypeVar('_ACM', bound=AsyncContextManager)
+_ACM = TypeVar('_ACM', bound=AsyncContextManager, covariant=True)

 class _ACMWrapper(Generic[_ACM]):
     """Async context manager wrapper"""
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 88ab38a..55a75b8 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -4168,7 +4168,7 @@ class SFTPClient:
                    attrs: SFTPAttrs = SFTPAttrs(),
                    encoding: Optional[str] = 'utf-8', errors: str = 'strict',
                    block_size: int = SFTP_BLOCK_SIZE,
-                   max_requests: int = _MAX_SFTP_REQUESTS) -> SFTPFileProtocol:
+                   max_requests: int = _MAX_SFTP_REQUESTS) -> SFTPClientFile:
         """Open a remote file

            This method opens a remote file and returns an
@@ -4300,8 +4300,7 @@ class SFTPClient:
                      attrs: SFTPAttrs = SFTPAttrs(),
                      encoding: Optional[str] = 'utf-8', errors: str = 'strict',
                      block_size: int = SFTP_BLOCK_SIZE,
-                     max_requests: int = _MAX_SFTP_REQUESTS) -> \
-            SFTPFileProtocol:
+                     max_requests: int = _MAX_SFTP_REQUESTS) -> SFTPClientFile:
         """Open a remote file using SFTP v5/v6 flags

            This method is very similar to :meth:`open`, but the pflags_or_mode
```

This change is now in the "develop" branch as commit a4b2615. Thanks for reporting this!

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.10.0.
